From 129bd51379857423b41cbd8b35e60960c30dcaac Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Wed, 10 Jul 2024 16:45:54 +0800
Subject: [PATCH 83/98] libinput-seat: Avoid processing other seats in udev

Maintain a separate list for udev-managed physical seats and only
process those, preventing other seats from interfering with physical
input device handling.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 libweston/libinput-seat.c | 11 ++++++++---
 libweston/libinput-seat.h |  2 ++
 2 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/libweston/libinput-seat.c b/libweston/libinput-seat.c
index 9e820f60..8a48adaf 100644
--- a/libweston/libinput-seat.c
+++ b/libweston/libinput-seat.c
@@ -347,7 +347,7 @@ udev_input_enable(struct udev_input *input)
 		process_events(input);
 	}
 
-	wl_list_for_each(seat, &input->compositor->seat_list, base.link) {
+	wl_list_for_each(seat, &input->seat_list, link) {
 		evdev_notify_keyboard_focus(&seat->base, &seat->devices_list);
 
 		if (!wl_list_empty(&seat->devices_list))
@@ -392,6 +392,8 @@ udev_input_init(struct udev_input *input, struct weston_compositor *c,
 
 	memset(input, 0, sizeof *input);
 
+	wl_list_init(&input->seat_list);
+
 	input->compositor = c;
 	input->configure_device = configure_device;
 	if (c->disable_input) {
@@ -438,7 +440,7 @@ udev_input_destroy(struct udev_input *input)
 
 	if (input->libinput_source)
 		wl_event_source_remove(input->libinput_source);
-	wl_list_for_each_safe(seat, next, &input->compositor->seat_list, base.link)
+	wl_list_for_each_safe(seat, next, &input->seat_list, link)
 		udev_seat_destroy(seat);
 	libinput_unref(input->libinput);
 }
@@ -516,6 +518,8 @@ udev_seat_create(struct udev_input *input, const char *seat_name)
 
 	wl_list_init(&seat->devices_list);
 
+	wl_list_insert(input->seat_list.prev, &seat->link);
+
 	return seat;
 }
 
@@ -534,6 +538,7 @@ udev_seat_destroy(struct udev_seat *seat)
 	wl_list_remove(&seat->output_destroyed_listener.link);
 	wl_list_remove(&seat->output_moved_listener.link);
 	wl_list_remove(&seat->output_heads_listener.link);
+	wl_list_remove(&seat->link);
 	free(seat);
 }
 
@@ -542,7 +547,7 @@ udev_seat_get_named(struct udev_input *input, const char *seat_name)
 {
 	struct udev_seat *seat;
 
-	wl_list_for_each(seat, &input->compositor->seat_list, base.link) {
+	wl_list_for_each(seat, &input->seat_list, link) {
 		if (strcmp(seat->base.seat_name, seat_name) == 0)
 			return seat;
 	}
diff --git a/libweston/libinput-seat.h b/libweston/libinput-seat.h
index 636636fb..10d97a42 100644
--- a/libweston/libinput-seat.h
+++ b/libweston/libinput-seat.h
@@ -43,6 +43,7 @@ struct udev_seat {
 	struct wl_listener output_destroyed_listener;
 	struct wl_listener output_moved_listener;
 	struct wl_listener output_heads_listener;
+	struct wl_list link;
 };
 
 typedef void (*udev_configure_device_t)(struct weston_compositor *compositor,
@@ -54,6 +55,7 @@ struct udev_input {
 	struct weston_compositor *compositor;
 	int suspended;
 	udev_configure_device_t configure_device;
+	struct wl_list seat_list;
 };
 
 int
-- 
2.39.5

